src/CodeStyle/Core/CodeFixes/AbstractFormattingAnalyzerImpl.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    internal abstract class AbstractFormattingAnalyzerImpl
+    {
+        private readonly DiagnosticDescriptor _descriptor;
+
+        protected AbstractFormattingAnalyzerImpl(DiagnosticDescriptor descriptor)
+        {
+            _descriptor = descriptor;
+        }
+
+        internal void InitializeWorker(AnalysisContext context)
+        {
+            var workspace = new AdhocWorkspace();
+            var codingConventionsManager = CodingConventionsManagerFactory.CreateCodingConventionsManager();
+
+            context.RegisterSyntaxTreeAction(c => AnalyzeSyntaxTree(c, workspace, codingConventionsManager));
+        }
+
+        protected abstract OptionSet ApplyFormattingOptions(OptionSet optionSet, ICodingConventionContext codingConventionContext);
+
+        private void AnalyzeSyntaxTree(SyntaxTreeAnalysisContext context, Workspace workspace, ICodingConventionsManager codingConventionsManager)
+        {
+            var tree = context.Tree;
+            var cancellationToken = context.CancellationToken;
+
+            var options = workspace.Options;
+            if (File.Exists(tree.FilePath))
+            {
+                var codingConventionContext = codingConventionsManager.GetConventionContextAsync(tree.FilePath, cancellationToken).GetAwaiter().GetResult();
+                options = ApplyFormattingOptions(options, codingConventionContext);
+            }
+
+            var oldText = tree.GetText(cancellationToken);
+            var formattingChanges = Formatter.GetFormattedTextChanges(tree.GetRoot(cancellationToken), workspace, options, cancellationToken);
+
+            // formattingChanges could include changes that impact a larger section of the original document than
+            // necessary. Before reporting diagnostics, process the changes to minimize the span of individual
+            // diagnostics.
+            foreach (var formattingChange in formattingChanges)
+            {
+                var change = formattingChange;
+                if (change.NewText.Length > 0 && !change.Span.IsEmpty)
```

Comment:
why the .IsEmpty check?

Reply:
When `IsEmpty` is true, the change is an insertion so the reported span is already the minimum possible span (a position).